nsqd/topic.go

```diff
@@ -329,6 +385,19 @@ func (t *Topic) messagePump() {
 					t.name, msg.ID, channel.name, err)
 			}
 		}
+
+		// If in draining mode and we wrote a message to channels
+		// check if it was the last message on the topic (there are no more left)
+		// in which case we start draining each channel
+		if atomic.LoadInt32(&t.isDraining) == 1 {
```

Comment:
Probably worth benchmarking... but we've tried hard to minimize the logic in the core message loop. In this case, draining is an irreversible state, meaning we just need to propagate the state change locally into this goroutine (e.g. like paused state) rather than check an atomic for every message published.

Reply:
I tried to write some benchmarks generally for `messagePump` but I couldn't get a consistent benchmark output. (as such i'll probably drop that benchmark from this PR before landing - unless it looks useful). In trying to benchmark i ran into issues with GUID sequences expiring so I switched some benchmarks to use an integer sequence to avoid timing impacts from GUID sequences expiring.

```
name                old time/op   new time/op   delta
TopicMessagePump-2  3.06µs ±181%  7.04µs ±131%   ~     (p=0.113 n=9+10)
```

I'd assume a refactor which added another 'drainChan` channel for `Topic.messagePump` to watch for draining state updates would be more expensive than `atomic.LoadInt32` but.. i'm not sure - @ploxiln have any intuition here?